### PR TITLE
[funimation] Add new extractor

### PR DESCRIPTION
**Pull request by Slyneth**

### CODE REVIEW

**Comment by Slyneth** [comment]
I messed up with the git while trying to squash. I have no idea where that *Merge remote-tracking branch 'rg3/master'* came from or why my commit got deleted.

**Comment by Slyneth** [comment]
This is finally ready, I hope.

File: youtube_dl/extractor/funimation.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            login = self._download_webpage(login_request, login_url)
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 403:
+                raise ExtractorError('Funimation is not available in your region.', expected=True)
+            raise
+        if re.search(r'<meta property="og:url" content="http://www.funimation.com/login"/>', login) is not None:
+                raise ExtractorError('Unable to login, wrong username or password.', expected=True)
+
+    def _real_extract(self, url):
+        self._login()
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: Login should be in `_real_initialize`.

File: youtube_dl/extractor/funimation.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+from ..utils import (
+    encode_dict,
+    sanitized_Request,
+    ExtractorError,
+    urlencode_postdata
+)
+import re
+
+class FunimationIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?funimation\.com/shows/.*/videos/official/(?P<id>[^?]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: It should not match URLs like http://www.funimation.com/shows//videos/official/breeze.

File: youtube_dl/extractor/funimation.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        try:
+            webpage = self._download_webpage(url, video_id)
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 403:
+                raise ExtractorError('Funimation is not available in your region.', expected=True)
+            raise
+        if re.search(r'"sdUrl":"http', webpage) is None:
+                raise ExtractorError('You are not logged-in or the stream requires subscription.', expected=True)
+
+        m3u8 = self._search_regex(r'"(?:.+)Url":"(.+?m3u8)"', webpage, 'm3u8') + self._search_regex(r'"authToken":"(.+?)"', webpage, 'm3u8')
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: What is the purpose of the first group in regex?

File: youtube_dl/extractor/funimation.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        formats = self._extract_m3u8_formats(m3u8.replace('\\', ''), video_id, ext='mp4', entry_protocol='m3u8_native')
+
+        video_title = self._og_search_title(webpage)
+        if not video_title:
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: Pass `default` to `_og_search_title` instead.

File: youtube_dl/extractor/funimation.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            'password_field': password,
+        }))
+        login_request = sanitized_Request(login_url, data)
+        login_request.add_header('Content-Type', 'application/x-www-form-urlencoded')
+        try:
+            login = self._download_webpage(login_request, login_url)
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 403:
+                raise ExtractorError('Funimation is not available in your region.', expected=True)
+            raise
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: Code at lines 37-42 is similar to code @ 51-56. It should be moved to separate routine.

File: youtube_dl/extractor/funimation.py (hunk 3 above)
Comment (Slyneth) [inline youtube_dl/extractor/funimation.py]: I am still learning python regex, my mistake to create a group.

File: youtube_dl/extractor/funimation.py (hunk 4 above)
Comment (Slyneth) [inline youtube_dl/extractor/funimation.py]: I don't understand this. I have been told to create a fallback title, this is why I did it like this.

File: youtube_dl/extractor/funimation.py (hunk 5 above)
Comment (Slyneth) [inline youtube_dl/extractor/funimation.py]: I am not sure I understand enough to the that. Could you direct me to an example?

Comment (Slyneth) [inline youtube_dl/extractor/funimation.py]: Could I maybe put a seperate geographic restriction check in _real_initialize, before self._login?

**Comment by dstftw** [comment]
Newly created funimation accounts does not seem to work at all. Can you provide account credentials for test?

**Comment by Slyneth** [comment]
With the current commit, I have used my new account with a FreePass. It worked with the one in the test. Most of the videos doesn't work with a free account as they are hosted with Hulu, but there are some that works, like the one in the test. But I spend an hour trying to find it, they are very rare. Funimation is very useless without a paid subscription.

My SubPass still works as intended. Here is an example:

    python ~/Dev/youtube-dl/youtube_dl/__main__.py --username ytdl@muratcansimsek.com --password ytdltest http://www.funimation.com/shows/air/videos/official/breeze          
    [Funimation] http://www.funimation.com/login: Downloading webpage
    [Funimation] breeze: Downloading webpage
    [Funimation] breeze: Downloading m3u8 information
    [hlsnative] Downloading m3u8 manifest
    [hlsnative] Total fragments: 146
    [download] Destination: Air - Breeze-AIRENG0001.mp4
    [download]   0.0% of ~272.29MiB at 171.36KiB/s ETA 41:27

I also created a trial account. I can't give credentials for it as you can't use it without credit card information but it appears it works too:

    python ~/Dev/youtube-dl/youtube_dl/__main__.py http://www.funimation.com/shows/noragami/videos/official/wish?watch=dub
    [Funimation] http://www.funimation.com/login: Downloading webpage
    [Funimation] wish: Downloading webpage
    [Funimation] wish: Downloading m3u8 information
    [hlsnative] Downloading m3u8 manifest
    [hlsnative] Total fragments: 147
    [download] Destination: Noragami - Wish-NRGENG6m9gJM0016.mp4
    [download]   0.0% of ~1.23GiB at 94.58KiB/s ETA 04:36:38

Can you post the error you get?

File: youtube_dl/extractor/funimation.py (hunk 5 above)
Comment (Slyneth) [inline youtube_dl/extractor/funimation.py]: I tried, I can't manage this one.

**Comment by dstftw** [comment]
Cherry-picked in ab4bdc913fafc71217c1196602525e0382c4eea3, thanks.

**Pull request closed without merging**